### PR TITLE
feat: extract flows and templates

### PR DESCRIPTION
**Pull request by loicmathieu**
Allow extracting Flows and Templates

### CODE REVIEW

**Comment by Skraye** [comment]
You may follow the same pattern as bulk actions on executions to keep consistency and easier implementation in the front 
* [By query](https://github.com/kestra-io/kestra/blob/d12ea0da4e51aa1544d11b65e77fe8a82f3d0a93/webserver/src/main/java/io/kestra/webserver/controllers/ExecutionController.java#L327)
* [By Ids](https://github.com/kestra-io/kestra/blob/d12ea0da4e51aa1544d11b65e77fe8a82f3d0a93/webserver/src/main/java/io/kestra/webserver/controllers/ExecutionController.java#L288)

**Comment by tchiotludo** [comment]
Totally agree, `/extract/by-ids` and `/extract/by-query` is better I think

File: cli/src/main/java/io/kestra/cli/commands/flows/namespaces/FlowNamespaceExtractCommand.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    @CommandLine.Parameters(index = "1", description = "the directory to extract the file to")
+    public Path directory;
+
+    @Override
+    public Integer call() throws Exception {
+        super.call();
+
+        try(DefaultHttpClient client = client()) {
+            MutableHttpRequest<Object> request = HttpRequest
+                .GET("/api/v1/flows/extract/by_query?namespace=" + namespace).accept(MediaType.APPLICATION_OCTET_STREAM);
```
Comment (tchiotludo) [inline cli/src/main/java/io/kestra/cli/commands/flows/namespaces/FlowNamespaceExtractCommand.java:38]:
```suggestion
                .GET("/api/v1/flows/extract/query?namespace=" + namespace).accept(MediaType.APPLICATION_OCTET_STREAM);
```

File: core/src/main/java/io/kestra/core/models/templates/Template.java [hunk 2]
```diff
@@ -25,6 +31,16 @@
 @ToString
 @EqualsAndHashCode
 public class Template implements DeletedInterface {
+
+    private static final ObjectMapper jsonMapper = JacksonMapper.ofJson().copy()
```
Comment (tchiotludo) [inline core/src/main/java/io/kestra/core/models/templates/Template.java:35]: Too many duplicate from flow

File: core/src/main/java/io/kestra/core/models/templates/Template.java [hunk 3]
```diff
@@ -93,6 +109,23 @@ public Optional<ConstraintViolationException> validateUpdate(Template updated) {
         }
     }
 
+    public String generateSource() {
```
Comment (tchiotludo) [inline core/src/main/java/io/kestra/core/models/templates/Template.java:112]: this method is not necessary, a simple mapper will generate the source on templates

File: webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java [hunk 4]
```diff
@@ -416,4 +421,59 @@ public List<ValidateConstraintViolation> validateFlows(
             })
             .collect(Collectors.toList());
     }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/extract/by_query", produces = MediaType.APPLICATION_OCTET_STREAM)
```
Comment (tchiotludo) [inline webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java:426]:
```suggestion
    @Get(uri = "/extract/query", produces = MediaType.APPLICATION_OCTET_STREAM)
```

File: webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java [hunk 5]
```diff
@@ -416,4 +421,59 @@ public List<ValidateConstraintViolation> validateFlows(
...
+            }
+
+            archive.finish();
+            return HttpResponse.ok(bos.toByteArray())
+                .header("Content-Disposition", "attachment; filename=\"flows.zip\"");
+        }
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Post(uri = "/extract/by_ids", produces = MediaType.APPLICATION_OCTET_STREAM, consumes = MediaType.APPLICATION_JSON)
```
Comment (tchiotludo) [inline webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java:454]:
```suggestion
    @Post(uri = "/extract/by-ids", produces = MediaType.APPLICATION_OCTET_STREAM, consumes = MediaType.APPLICATION_JSON)
```

File: webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java [hunk 6]
```diff
@@ -416,4 +421,59 @@ public List<ValidateConstraintViolation> validateFlows(
...
+    public HttpResponse<byte[]> extractByQuery(
+        @Parameter(description = "A string filter") @Nullable @QueryValue(value = "q") String query,
+        @Parameter(description = "A namespace filter prefix") @Nullable @QueryValue(value = "namespace") String namespace,
+        @Parameter(description = "A labels filter") @Nullable @QueryValue List<String> labels
+    ) throws IOException {
+        try(ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            ZipOutputStream archive = new ZipOutputStream(bos)) {
+
+            List<FlowWithSource> toExtract = flowRepository.findWithSource(query, namespace, RequestUtils.toMap(labels));
+            for(var flowWithSource : toExtract) {
```
Comment (tchiotludo) [inline webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java:440]: too many duplication, create a service that create the zip

**Review by tchiotludo — CHANGES_REQUESTED** [review]

File: cli/src/main/java/io/kestra/cli/commands/flows/namespaces/FlowNamespaceExtractCommand.java (hunk 1 above)
Comment (loicmathieu) [reply cli/src/main/java/io/kestra/cli/commands/flows/namespaces/FlowNamespaceExtractCommand.java:38]: Well, see your comment bellow

> Totally agree, /extract/by-ids and /extract/by-query is better I think

I'm OK to change it but as extract is a verb by_query reads better than just query.

Comment (loicmathieu) [reply cli/src/main/java/io/kestra/cli/commands/flows/namespaces/FlowNamespaceExtractCommand.java:38]: Moreover, if we keep `by-ids` we should use `by-query`

File: core/src/main/java/io/kestra/core/models/templates/Template.java (hunk 3 above)
Comment (loicmathieu) [reply core/src/main/java/io/kestra/core/models/templates/Template.java:113]: Yes, I setup the object mapper once for all and use it directly.
By the way, I think the same can be done on the flow class, as the `Flow.generateSource()` method is now only used in tests this should be fine to change it the same way I did for templates.

File: webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java (hunk 5 above)
Comment (loicmathieu) [reply webserver/src/main/java/io/kestra/webserver/controllers/FlowController.java]: done

**Pull request merged**